Add docs for using Auth claims within modules (#3545)
clockworklabs/SpacetimeDB: docs/docs/11-How-To/03-using-auth-claims.md

---
slug: /how-to/using-auth-claims
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Using Auth Claims

SpacetimeDB allows you to easily access authentication (auth) claims embedded in OIDC-compliant JWT tokens. Auth claims are key-value pairs that provide information about the authenticated user. For example, they may contain a user's unique ID, email, or authentication provider. If you want to view these fields for yourself, you can inspect the contents of any JWT using online tools like [jwt.io](https://jwt.io/).

Within a SpacetimeDB reducer, you can access the auth claims from a client's token via the `ReducerContext`. Below are some examples of how to use these claims.

## Accessing Common Claims: Subject and Issuer

> The subject (`sub`) and issuer (`iss`) are the most commonly accessed claims in a JWT. The subject usually represents the user's unique identifier, while the issuer indicates which authentication provider issued the token. These are required claims, which are used to compute each user's `Identity`. Because these are so commonly used, there are helper functions to get them.

Reviewer: ```suggestion
The subject (`sub`) and issuer (`iss`) are the most commonly accessed claims in a JWT. The issuer indicates which authentication provider issued the token, and the subject represents the unique identifier assigned to the user by the issuer. These are required claims, which are used to compute each user's `Identity`. Because these are so commonly used, there are helper functions to get them.
```

<Tabs groupId="server-language" defaultValue="typescript">
<TabItem value="rust" label="Rust">

```rust
#[reducer(client_connected)]
pub fn connect(ctx: &ReducerContext) -> Result<(), String> {
    let auth_ctx = ctx.sender_auth();
    let (subject, issuer) = match auth_ctx.jwt() {
        Some(claims) => (claims.subject().to_string(), claims.issuer().to_string()),
        None => {
            return Err("Client connected without JWT".to_string());
        }
    };
    log::info!("sub: {}, iss: {}", subject, issuer);
    Ok(())
}
```

*Example output when using a Google-issued token:*
```
INFO: src\lib.rs:64: sub: 321321321321321, iss: https://accounts.google.com

> ```

Reviewer: Is there a reason this is just in the Rust one?

</TabItem>
<TabItem value="csharp" label="C#">

```cs
[Reducer(ReducerKind.ClientConnected)]
public static void ClientConnected(ReducerContext ctx)
{
    var claims = ctx.SenderAuth.Jwt ?? throw new Exception("Client connected without JWT");
    Log.Info($"Client connected with csub: {claims.Subject}, and iss: {claims.Issuer}");
}
```

</TabItem>
<TabItem value="typescript" label="TS">

```typescript
import { SenderError } from "spacetimedb/server";

spacetimedb.clientConnected((ctx) => {
  const jwt = ctx.senderAuth.jwt;
  if (jwt == null) {
    throw new SenderError("Unauthorized: JWT is required to connect");
  }
  console.info(`Client connected with sub: ${jwt.subject}, iss: ${jwt.issuer}`);
});
```

</TabItem>
</Tabs>

## Example: Restricting auth providers

> Since users can use any valid token to connect to SpacetimeDB, their token may originate from any authentication provider. For example, they could send an OIDC compliant token from Github even though you only want to accept tokens from Google. It is best practice to check at least the issuer and audience of tokens when clients connect, so you can ensure that your data can only be accessed by users of your application.

Reviewer: ```suggestion
Since users can use any valid token to connect to SpacetimeDB, their token may originate from any authentication provider. For example, they could send an OIDC compliant token from GitHub even though you only want to accept tokens from Google. It often makes sense to restrict access to your module to users issued by your own issuer or specific issuers. It is best practice to check at least the issuer when clients connect, so you can ensure that your data can only be accessed by users of your application.
```

Reviewer: I am still concerned about having users validate the `aud` claim. I do not believe we should ever deliver tokens to a module whose aud does not match the aud claim of the token.

Reviewer: I think for now we can leave this up to the module developer, but in 2.0 I would like to change this to be part of the database's configuration and be enforced by SpacetimeDB so these tokens never actually arrive at the module at all.
